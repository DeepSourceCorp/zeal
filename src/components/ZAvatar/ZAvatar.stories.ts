import ZAvatar from './ZAvatar.vue'

export default {
  title: 'Avatar',
  component: ZAvatar,
  excludeStories: /.*Data$/
}

export const AsALink = () => ({
  components: { ZAvatar },
  template: `
    <div class='padded-container'>
      <z-avatar
        type="link"
        image="https://randomuser.me/api/portraits/women/24.jpg"
        href="https://example.com"
      ></z-avatar>
    </div>
    `
})

export const InLoadingState = () => ({
  components: { ZAvatar },
  template: `
    <div class='padded-container'>
      <z-avatar
        :loading="true"
        type="link"
        image="https://randomuser.me/api/portraits/women/24.jpg"
        href="https://example.com"
      ></z-avatar>
    </div>
    `
})


export const WithUserName = () => ({
  components: { ZAvatar },
  template: `
    <div class='padded-container'>
      <z-avatar
        user-name="Akshay Paliwal"
      ></z-avatar>
    </div>
    `
})

export const WithImageAndUserName = () => ({
  components: { ZAvatar },
  template: `<div class='padded-container'>
      <z-avatar
        image="https://randomuser.me/api/portraits/women/24.jpg"
      ></z-avatar>
    </div>`
})

export const WithSizes = () => ({
  components: { ZAvatar },
  template: `
    <div class='bg-dark padded-container'>
      <div class='padded-container'>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="xs"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="sm"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="md"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="lg"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="xl"
        ></z-avatar>
      </div>
      <div class='padded-container bg-vanilla-100'>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="xs"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="sm"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="md"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="lg"
        ></z-avatar>
        <z-avatar
          image="https://randomuser.me/api/portraits/women/24.jpg"
          size="xl"
        ></z-avatar>
      </div>
      <div class='padded-container'>
        <z-avatar
          user-name="Akshay Paliwal"
          size="xs"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="sm"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="md"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="lg"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="xl"
        ></z-avatar>
      </div>
      <div class='bg-vanilla-100 padded-container'>
        <z-avatar
          user-name="Akshay Paliwal"
          size="xs"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="sm"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="md"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="lg"
        ></z-avatar>
        <z-avatar
          user-name="Akshay Paliwal"
          size="xl"
        ></z-avatar>
      </div>
    </div>
    `
})
