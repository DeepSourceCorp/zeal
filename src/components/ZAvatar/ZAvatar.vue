<template>
  <component v-if="isVisible" :is="TYPES[type]" class="inline-block rounded-full" :class="stroke">
    <span
      v-if="isLoading"
      :class="[`${SIZES[getSize].classes}`]"
      class="flex items-center justify-center text-center rounded-full bg-gradient-skeleton text-vanilla-300"
    >
    </span>
    <template v-else>
      <img
        v-if="this.image"
        :src="image"
        :alt="userName"
        class="rounded-full"
        :class="[`${SIZES[getSize].classes}`]"
        @error.once="setFallbackImage"
      />
      <span
        v-if="!this.image && this.userName"
        :class="[`${SIZES[getSize].classes}`]"
        class="flex items-center justify-center text-center rounded-full bg-ink-200 text-vanilla-300"
      >
        {{ getUserInitials }}
      </span>
    </template>
  </component>
</template>

<script lang="ts">
import Vue from 'vue'
const SIZES = {
  '2xs': { classes: 'leading-none h-3 w-3 text-xxs', text: '2xs' },
  xs: { classes: 'leading-none h-4 w-4 text-xxs', text: 'xs' },
  sm: { classes: 'leading-none h-6 w-6 text-xs', text: 'sm' },
  md: { classes: 'leading-none h-8 w-8 text-sm', text: 'md' },
  lg: { classes: 'leading-none h-12 w-12 text-lg', text: 'lg' },
  xl: { classes: 'leading-none h-16 w-16 text-2xl', text: 'xl' }
}

const TYPES = {
  link: 'a',
  div: 'div',
  span: 'span'
}

export default Vue.extend({
  name: 'ZAvatar',
  props: {
    type: {
      type: String,
      default: 'link',
      validator: (type) => Object.keys(TYPES).includes(type)
    },
    loading: {
      type: Boolean,
      default: false
    },
    image: String,
    fallbackImage: {
      type: String,
      required: false,
      default: ''
    },
    userName: {
      type: String,
      required: true
    },
    size: {
      type: String,
      default: SIZES.md.text,
      validator: (size) => Object.keys(SIZES).includes(size)
    },
    stroke: {
      type: String,
      required: false,
      default: 'bg-ink-300 p-0.5'
    }
  },
  data() {
    return {
      isVisible: true,
      SIZES,
      TYPES
    }
  },
  computed: {
    isLoading(): boolean {
      const { $options, $props } = this.$parent as Vue
      return $options.name === 'ZAvatarGroup' ? $props.loading : this.loading
    },
    getSize(): string {
      /**
       * Returns `size` property as size of avatar if parent is not ZAvatarGroup,
       * else it returns `size` property of parent as size of avatar.
       */
      const { $options, $props } = this.$parent as Vue
      return $options.name === 'ZAvatarGroup' ? $props.size : this.size
    },
    getUserInitials(): string {
      /**
       * Returns capitalized initials of the user.
       * (Only from the first two words of the userName)
       */
      return this.userName
        ? this.userName
            .split(' ')
            .map((elem) => elem.charAt(0))
            .slice(0, this.size === '2xs' ? 1 : 2) // Make it a single alphabet for the `2xs` size variant
            .join('')
            .toUpperCase()
        : ''
    }
  },
  methods: {
    setFallbackImage(event: ErrorEvent): void {
      if (!this.fallbackImage) return

      const el = event.target as HTMLImageElement
      if (el.src !== this.fallbackImage) {
        el.src = this.fallbackImage
      }
    }
  }
})
</script>
