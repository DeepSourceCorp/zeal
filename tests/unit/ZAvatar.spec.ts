import Vue from 'vue'
import { createLocalVue, mount, Wrapper } from '@vue/test-utils'

import ZAvatar from '../../src/components/ZAvatar'

describe('Avatar component', () => {
  let mountFunction: (options?: Record<string, unknown>) => Wrapper<Vue>
  let localVue: typeof Vue
  beforeEach(() => {
    localVue = createLocalVue()
    mountFunction = (options = {}) => {
      return mount(ZAvatar, {
        localVue,
        ...options
      })
    }
  })
  it('renders an avatar when the component is mounted', () => {
    expect(
      mountFunction({
        propsData: {
          userName: 'John Doe'
        }
      }).html()
    ).toMatchSnapshot()
  })

  it('renders root element as a link for avatar with no props', () => {
    let mountedInstance = mountFunction({
      propsData: {
        userName: 'John Doe'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.find('a').exists()).toBe(true)
  })

  it('renders root element as a span for avatar with type prop `span`', () => {
    let mountedInstance = mountFunction({
      propsData: {
        type: 'span',
        userName: 'John Doe'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.find('span').exists()).toBe(true)
  })

  it('renders avatar with skeleton gradient background when loading is true', () => {
    let mountedInstance = mountFunction({
      propsData: {
        loading: true,
        userName: 'John Doe'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.find('.bg-gradient-skeleton').exists()).toBe(true)
  })

  it('renders avatar correctly, when image is not given', () => {
    let mountedInstance = mountFunction({
      propsData: {
        userName: 'John Doe'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    const spanElement = mountedInstance.find('span.rounded-full')
    expect(spanElement.exists()).toBe(true)
    expect(spanElement.text()).toContain('JD')
  })

  it('renders avatar correctly, when image is given', () => {
    let mountedInstance = mountFunction({
      propsData: {
        userName: 'John Doe',
        image: 'https://example.com/some.jpg'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    const imageElement = mountedInstance.find('img.rounded-full')
    expect(imageElement.exists()).toBe(true)
    expect(imageElement.attributes().src).toEqual('https://example.com/some.jpg')
    expect(imageElement.attributes().alt).toEqual('John Doe')
  })

  it('renders fallback image', () => {
    let mountedInstance = mountFunction({
      propsData: {
        userName: 'John Doe',
        image: 'https://example.com/some.jpg',
        fallbackImage: 'https://example.com/fallback.jpg'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    const imageElement = mountedInstance.find('img.rounded-full')
    imageElement.trigger('error')
    expect(imageElement.exists()).toBe(true)
    expect(imageElement.attributes().src).toEqual('https://example.com/fallback.jpg')
    expect(imageElement.attributes().alt).toEqual('John Doe')
  })

  it('renders avatar correctly, when size is given', () => {
    let mountedInstance = mountFunction({
      propsData: {
        userName: 'John Doe',
        size: 'sm'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    const spanElement = mountedInstance.find('span.rounded-full')
    expect(spanElement.exists()).toBe(true)
    expect(spanElement.attributes().class).toContain('leading-none h-6 w-6 text-xs')
  })
})
