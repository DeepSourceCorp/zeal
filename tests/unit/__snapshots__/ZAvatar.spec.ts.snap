// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component renders an avatar when the component is mounted 1`] = `
<a class="inline-block rounded-full bg-ink-300 p-0.5">
  <!----> <span class="flex items-center justify-center text-center rounded-full bg-ink-200 text-vanilla-300 leading-none h-8 w-8 text-sm">
      JD
    </span>
</a>
`;

exports[`Avatar component renders avatar correctly, when image is given 1`] = `
<a class="inline-block rounded-full bg-ink-300 p-0.5"><img src="https://example.com/some.jpg" alt="John Doe" class="rounded-full leading-none h-8 w-8 text-sm">
  <!---->
</a>
`;

exports[`Avatar component renders avatar correctly, when image is not given 1`] = `
<a class="inline-block rounded-full bg-ink-300 p-0.5">
  <!----> <span class="flex items-center justify-center text-center rounded-full bg-ink-200 text-vanilla-300 leading-none h-8 w-8 text-sm">
      JD
    </span>
</a>
`;

exports[`Avatar component renders avatar correctly, when size is given 1`] = `
<a class="inline-block rounded-full bg-ink-300 p-0.5">
  <!----> <span class="flex items-center justify-center text-center rounded-full bg-ink-200 text-vanilla-300 leading-none h-6 w-6 text-xs">
      JD
    </span>
</a>
`;

exports[`Avatar component renders avatar with skeleton gradient background when loading is true 1`] = `<a class="inline-block rounded-full bg-ink-300 p-0.5"><span class="flex items-center justify-center text-center rounded-full bg-gradient-skeleton text-vanilla-300 leading-none h-8 w-8 text-sm"></span></a>`;

exports[`Avatar component renders fallback image 1`] = `
<a class="inline-block rounded-full bg-ink-300 p-0.5"><img src="https://example.com/some.jpg" alt="John Doe" class="rounded-full leading-none h-8 w-8 text-sm">
  <!---->
</a>
`;

exports[`Avatar component renders root element as a link for avatar with no props 1`] = `
<a class="inline-block rounded-full bg-ink-300 p-0.5">
  <!----> <span class="flex items-center justify-center text-center rounded-full bg-ink-200 text-vanilla-300 leading-none h-8 w-8 text-sm">
      JD
    </span>
</a>
`;

exports[`Avatar component renders root element as a span for avatar with type prop \`span\` 1`] = `
<span class="inline-block rounded-full bg-ink-300 p-0.5"><!----> <span class="flex items-center justify-center text-center rounded-full bg-ink-200 text-vanilla-300 leading-none h-8 w-8 text-sm">
      JD
    </span></span>
`;
